### ISSUE TITLE
Possibly incorrect branch distance calculation

### ISSUE THREAD

**Issue by MarekPokornyOva** [dotnet/corefx#26621]
Type: System.Reflection.Metadata.Ecma335.ControlFlowBuilder (in https://github.com/dotnet/corefx/blob/master/src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/ControlFlowBuilder.cs)

```
internal bool IsShortBranchDistance(ImmutableArray<int>.Builder labels, out int distance)
{
    const int shortBranchSize = 2;
    const int longBranchSize = 5;

    int labelTargetOffset = labels[Label.Id - 1];

    if (labelTargetOffset < 0)
    {
        Throw.InvalidOperation_LabelNotMarked(Label.Id);
    }

//POSSIBLE ISSUE SPOT!!!
    distance = labelTargetOffset - (ILOffset + shortBranchSize);

    if (unchecked((sbyte)distance) == distance)
    {
        return true;
    }

    distance = labelTargetOffset - (ILOffset + longBranchSize);
    return false;
}
```
Do I understand correctly, this doesn't follow really written instruction (leave / leave.s) and presume a compiler uses only short instruction for short jump? What happens if the compiler uses "expensive" long instruction even for short jumps?
--------------
Consider this compiler:
```
StandaloneSignatureHandle MainGenerator(InstructionEncoder gen)
{
    BlobBuilder builder = new BlobBuilder();
    LocalVariablesEncoder encoder = new BlobEncoder(builder).LocalVariableSignature(2);
    BuildSignature(encoder.AddVariable().Type(), typeof(int));
    BuildSignature(encoder.AddVariable().Type(), typeof(Exception));
    StandaloneSignatureHandle localVariablesSignature = _tablesAndHeaps.AddStandaloneSignature(_tablesAndHeaps.GetOrAddBlob(builder));

    LabelHandle tryStart = gen.DefineLabel();
    LabelHandle tryEnd = gen.DefineLabel();
    LabelHandle catchStart = gen.DefineLabel();
    LabelHandle catchEnd = gen.DefineLabel();

    gen.ControlFlowBuilder.AddCatchRegion(tryStart, tryEnd, catchStart, catchEnd, GetTypeReference(typeof(Exception)));

    gen.OpCode(ILOpCode.Ldc_i4_0);
    gen.StoreLocal(0);

    gen.MarkLabel(tryStart);
    gen.LoadLocal(0);
    gen.LoadLocal(0);
    gen.OpCode(ILOpCode.Div);
    gen.StoreLocal(0);
    gen.Branch(ILOpCode.Leave, catchEnd);
    gen.MarkLabel(tryEnd);
    gen.MarkLabel(catchStart);
    gen.StoreLocal(1);
    gen.Branch(ILOpCode.Leave, catchEnd);
    gen.MarkLabel(catchEnd);

    gen.OpCode(ILOpCode.Ret);

    return localVariablesSignature;
}
```
Which generates this IL sequence:
0x0000 Ldc_i4_0
0x0001 Stloc_0
0x0002 Ldloc_0
0x0003 Ldloc_0
0x0004 Div
0x0005 Stloc_0
0x0006 Leave     09 (shouldn't be 6?)
0x000b Stloc_1
0x000c Leave     03 (shouldn't be 0?)
0x0011 Ret
-----------------------
I found out, such compiled code makes error during viewing in ILSpy (what is minor issue) but it also breaks loading the output assembly in AppDomain.

**Comment by jkotas** [comment]
cc @tmat

**Comment by tmat** [comment]
Fixed by  #28493
